### ISSUE TITLE
[Bug] RuntimeException Expected attribute count to match attribute storage count

### ISSUE THREAD

**Issue by butschster** [vimeo/psalm#7751]
Hello everybody.

Our project has been updated up to 4.22 version and after that psalm started crashing while project analyzing with the following error:

```
Uncaught Exception: RuntimeException Expected attribute count to match attribute storage count
Emitted in /Users/butschster/Webserver/spiral/framework/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/AttributesAnalyzer.php:117
Stack trace in the forked worker:
#0 /Users/butschster/Webserver/spiral/framework/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/ClassAnalyzer.php(406): Psalm\Internal\Analyzer\AttributesAnalyzer::analyze(Object(Psalm\Internal\Analyzer\ClassAnalyzer), Object(Psalm\Context), Object(Psalm\Storage\ClassLikeStorage), Array, 1, Array)
#1 /Users/butschster/Webserver/spiral/framework/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/StatementsAnalyzer.php(591): Psalm\Internal\Analyzer\ClassAnalyzer->analyze(Object(Psalm\Context), NULL)
#2 /Users/butschster/Webserver/spiral/framework/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/StatementsAnalyzer.php(206): Psalm\Internal\Analyzer\StatementsAnalyzer::analyzeStatement(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Stmt\Class_), Object(Psalm\Context), NULL)
#3 /Users/butschster/Webserver/spiral/framework/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/Statements/Block/IfElse/IfAnalyzer.php(68): Psalm\Internal\Analyzer\StatementsAnalyzer->analyze(Array, Object(Psalm\Context))
#4 /Users/butschster/Webserver/spiral/framework/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/Statements/Block/IfElseAnalyzer.php(365): Psalm\Internal\Analyzer\Statements\Block\IfElse\IfAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Stmt\If_), Object(Psalm\Internal\Scope\IfScope), Object(Psalm\Internal\Scope\IfConditionalScope), Object(Psalm\Context), Object(Psalm\Context), Object(Psalm\Context), Array)
#5 /Users/butschster/Webserver/spiral/framework/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/StatementsAnalyzer.php(517): Psalm\Internal\Analyzer\Statements\Block\IfElseAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Stmt\If_), Object(Psalm\Context))
#6 /Users/butschster/Webserver/spiral/framework/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/StatementsAnalyzer.php(206): Psalm\Internal\Analyzer\StatementsAnalyzer::analyzeStatement(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Stmt\If_), Object(Psalm\Context), NULL)
#7 /Users/butschster/Webserver/spiral/framework/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/NamespaceAnalyzer.php(92): Psalm\Internal\Analyzer\StatementsAnalyzer->analyze(Array, Object(Psalm\Context), NULL, true)
#8 /Users/butschster/Webserver/spiral/framework/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/FileAnalyzer.php(316): Psalm\Internal\Analyzer\NamespaceAnalyzer->collectAnalyzableInformation()
#9 /Users/butschster/Webserver/spiral/framework/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/FileAnalyzer.php(193): Psalm\Internal\Analyzer\FileAnalyzer->populateCheckers(Array)
#10 /Users/butschster/Webserver/spiral/framework/vendor/vimeo/psalm/src/Psalm/Internal/Codebase/Analyzer.php(362): Psalm\Internal\Analyzer\FileAnalyzer->analyze()
#11 /Users/butschster/Webserver/spiral/framework/vendor/vimeo/psalm/src/Psalm/Internal/Fork/Pool.php(211): Psalm\Internal\Codebase\Analyzer->Psalm\Internal\Codebase\{closure}(166, '/Users/butschst...')
#12 /Users/butschster/Webserver/spiral/framework/vendor/vimeo/psalm/src/Psalm/Internal/Codebase/Analyzer.php(428): Psalm\Internal\Fork\Pool->__construct(Object(Psalm\Config), Array, Object(Closure), Object(Closure), Object(Closure), Object(Closure))
#13 /Users/butschster/Webserver/spiral/framework/vendor/vimeo/psalm/src/Psalm/Internal/Codebase/Analyzer.php(291): Psalm\Internal\Codebase\Analyzer->doAnalysis(Object(Psalm\Internal\Analyzer\ProjectAnalyzer), 3)
#14 /Users/butschster/Webserver/spiral/framework/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/ProjectAnalyzer.php(685): Psalm\Internal\Codebase\Analyzer->analyzeFiles(Object(Psalm\Internal\Analyzer\ProjectAnalyzer), 3, false, true)
#15 /Users/butschster/Webserver/spiral/framework/vendor/vimeo/psalm/src/Psalm/Internal/Cli/Psalm.php(373): Psalm\Internal\Analyzer\ProjectAnalyzer->check('/Users/butschst...', true)
#16 /Users/butschster/Webserver/spiral/framework/vendor/vimeo/psalm/psalm(7): Psalm\Internal\Cli\Psalm::run(Array)
#17 {main} in /Users/butschster/Webserver/spiral/framework/vendor/vimeo/psalm/src/Psalm/Internal/Fork/Pool.php:402
Stack trace:
#0 /Users/butschster/Webserver/spiral/framework/vendor/vimeo/psalm/src/Psalm/Internal/Fork/Pool.php(436): Psalm\Internal\Fork\Pool->readResultsFromChildren()
#1 /Users/butschster/Webserver/spiral/framework/vendor/vimeo/psalm/src/Psalm/Internal/Codebase/Analyzer.php(503): Psalm\Internal\Fork\Pool->wait()
#2 /Users/butschster/Webserver/spiral/framework/vendor/vimeo/psalm/src/Psalm/Internal/Codebase/Analyzer.php(291): Psalm\Internal\Codebase\Analyzer->doAnalysis(Object(Psalm\Internal\Analyzer\ProjectAnalyzer), 3)
#3 /Users/butschster/Webserver/spiral/framework/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/ProjectAnalyzer.php(685): Psalm\Internal\Codebase\Analyzer->analyzeFiles(Object(Psalm\Internal\Analyzer\ProjectAnalyzer), 3, false, true)
#4 /Users/butschster/Webserver/spiral/framework/vendor/vimeo/psalm/src/Psalm/Internal/Cli/Psalm.php(373): Psalm\Internal\Analyzer\ProjectAnalyzer->check('/Users/butschst...', true)
#5 /Users/butschster/Webserver/spiral/framework/vendor/vimeo/psalm/psalm(7): Psalm\Internal\Cli\Psalm::run(Array)
#6 {main}
(Psalm 4.22.0@fc2c6ab4d5fa5d644d8617089f012f3bb84b8703 crashed due to an uncaught Throwable)
```

You can also view the error here https://github.com/spiral/framework/runs/5412294207?check_suite_focus=true

**Comment by orklah** [comment]
can you run with --debug-by-line ?

This should give you the last line Psalm could analyze. If you could post a snippet of what's going on in the code around this line, we could see clearer in there

**Comment by butschster** [comment]
> can you run with --debug-by-line ?
> 
> This should give you the last line Psalm could analyze. If you could post a snippet of what's going on in the code around this line, we could see clearer in there

Hmm. I've just run psalm 3 times  and every analyzing crashed on different file.

**First run**
```
/Users/butschster/Webserver/spiral/framework/src/Stempler/src/Compiler/SourceMap.php:104
Uncaught Exception: RuntimeException Expected attribute count to match attribute storage count
```

**Second run**
```
/Users/butschster/Webserver/spiral/framework/src/Attributes/src/polyfill.php:55
Uncaught Exception: RuntimeException Expected attribute count to match attribute storage count
```

**Third run**
```
/Users/butschster/Webserver/spiral/framework/src/Stempler/src/Lexer/Grammar/HTMLGrammar.php:109
Uncaught Exception: RuntimeException Expected attribute count to match attribute storage count
```

**Comment by orklah** [comment]
That's... unexpected.

@AndrolGenhald do you have an idea on that?

**Comment by AndrolGenhald** [comment]
That's strange, I looked at SourceMap.php and I don't even see any attributes there... Am I looking at the wrong commit maybe?

I'll see if I can debug it locally sometime in the next few days, @butschster I see commit d8ca6bec9ce307b97ab01a0b985fa38557344335 referenced in the GitHub action, I assume that's the one that's failing that I should debug?

**Comment by AndrolGenhald** [comment]
Ok, I've figured out the cause, but I'm not sure what to do about it yet. The problem is that we've got attributes coming from multiple stubs during scanning, which means we get `AttributeStorage`s for attributes declared from multiple stubs, but then when analyzing it's analyzing a single declaration that doesn't have all of the attributes expected.

The first one I ran into when debugging was actually for `Attribute` itself. We have `resources/stubs/php/reflection/Attribute.phpstub` which has no attributes, `src/Attributes/src/polyfill.php` which uses `#[Attribute(...)]`, and `vendor/symfony/polyfill-php80/Resources/stubs/Attribute.php` which uses `#[Attribute(...)]`.

Adding the attribute to `resources/stubs/php/reflection/Attribute.phpstub` works around the problem and allows the analysis to complete successfully. To actually fix this I think I might have to rework how we store attributes, I'll try to find some time to work on it next week.
